### PR TITLE
Issue #1551: Fix FinalLocalVariableCheck false-negative

### PR DESCRIPTION
**Pull request by Vladlis**
#1722 
> 1) please make it as "An example of how to configure the check to validate variable definition is:"

Done.

>  2)  mismatch of documentation and getRequiredTokens(). please update all examples (javadoc, xdoc).

Corrected examples, is it OK now?

> 3) please create new class that will store all information about scope (Map + Deque).

Done.

> 4) Please create UT that show that variables in different but nested scopes are named the same and one of them final another is not final and issue is reported to required variable.

Done.
https://github.com/Vladlis/checkstyle/commit/85bf34603f574445610d535a1bc55e5d69fe6d5c#diff-0b3755657c714c3c6e8b488b4d29606aR215

> 5) please generate reports against project to see that new violations are correct , please do check all of them yourself.

I've corrected all the new violations on CS sources, changes are in a in a separate commit.

> 6)  getDefaultToekns() , get AcceptableTokes()
pelase move TokenTypes.VARIABLE_DEF to the end of list to let future developer easily see a difference between collections,

Done.

> 7) private boolean defAndInitInDifferentLoop(DetailAST ast1, DetailAST ast2)
Why only Loops are checked there, that should be processing for any block structure that create extra scope (scope is "{ }").

There is already processing for all scopes, but here we look for loops, because there one statement can be executed more than once.

### CODE REVIEW

**Comment by romani** [comment]
Travis is passed, please rebase your changes

**Comment by romani** [comment]
such a huge changes require regression testing - before/after report comparison and diff verification.

**Comment by romani** [comment]
> ast.getParent().getType() != TokenTypes.OBJBLOCK

please make a method from from this expression, it takes me a few minutes to understand that this is a check for fields. Logic should be read more quickly.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 1]
```diff
@@ -280,81 +350,65 @@ private static boolean isInLambda(DetailAST paramDef) {
...
         final DetailAST classOrMethodOfAst2 =
             findClassOrConstructorOrMethodInWhichItIsDefined(ast2);
-
-        final String identifierOfAst1 =
-            classOrMethodOfAst1.findFirstToken(TokenTypes.IDENT).getText();
-        final String identifierOfAst2 =
-            classOrMethodOfAst2.findFirstToken(TokenTypes.IDENT).getText();
-
-        return identifierOfAst1.equals(identifierOfAst2);
+        return classOrMethodOfAst1 == classOrMethodOfAst2;
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: Codacy Issue found: [Use equals() to compare object references.](https://www.codacy.com/app/checkstyle/checkstyle/file/1111847767/issues/source?bid=2193633&fileBranchId=2892126#l373)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 2]
```diff
@@ -280,81 +350,65 @@ private static boolean isInLambda(DetailAST paramDef) {
...
+        DetailAST loop1 = ast1.getParent();
+        while (loop1 != null && !isLoopAst(loop1.getType())) {
+            loop1 = loop1.getParent();
+        }
+        DetailAST loop2 = ast2.getParent();
+        while (loop2 != null && !isLoopAst(loop2.getType())) {
+            loop2 = loop2.getParent();
+        }
+        return loop1 == null && loop2 == null
+                || loop1 != null && loop1 == loop2;
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: Codacy Issue found: [Use equals() to compare object references.](https://www.codacy.com/app/checkstyle/checkstyle/file/1111847767/issues/source?bid=2193633&fileBranchId=2892126#l392)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java [hunk 3]
```diff
@@ -280,81 +350,65 @@ private static boolean isInLambda(DetailAST paramDef) {
...
+     * Check if both the variables are in the same loop.
+     * @param ast1 variable to compare.
+     * @param ast2 variable to compare.
+     * @return true if both the variables are in the same loop.
      */
-    private void insertVariable(DetailAST ast) {
-        final Map<String, DetailAST> state = scopeStack.peek();
-        final DetailAST astNode = ast.findFirstToken(TokenTypes.IDENT);
-        state.put(astNode.getText(), astNode);
+    private static boolean isInTheSameLoop(DetailAST ast1, DetailAST ast2) {
```
Comment (mkordas) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: Codacy Issue found: [The method isInTheSameLoop() has an NPath complexity of 27](https://www.codacy.com/app/checkstyle/checkstyle/file/1111847767/issues/source?bid=2193633&fileBranchId=2892126#l382)

**Comment by romani** [comment]
@Vladlis , please rebase 

> The method isInTheSameLoop() has an NPath complexity of 27

is not required to fix.

**Comment by romani** [comment]
@Vladlis ,

release 6.12 is planned to be at 31 of October.
it would very good to finish this PR before release. Do you have time to finish PR ?

**Comment by Vladlis** [comment]
@romani, OK, I'll find some time to fix the points you've mentioned.

**Comment by romani** [comment]
@mkordas , please disable automatic codereview, that distract contributors and there are a lot of false-positives most likely due to misconfiguration.

**Comment by romani** [comment]
Codecoverage is incomplete - https://travis-ci.org/checkstyle/checkstyle/jobs/88377557#L945

**Comment by mkordas** [comment]
>  there are a lot of false-positives most likely due to misconfiguration.
Can you show me some false-positive? I don't see any.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java]: I do not see a problem there.

**Comment by romani** [comment]
@mkordas ,

>    The method isInTheSameLoop() has an NPath complexity of 27
is not required to fix.

We already discussed this, allowed NPATH is 200 , http://checkstyle.sourceforge.net/config_metrics.html#NPathComplexity . Why we bother about 27 ?

**Comment by romani** [comment]
@Vladlis , code is ok, changes to chekstyle project is also OK.

I understand that such Check is very demanding, there will be too much new violations, we will not be able to validate them all.
Please share report before/after for senvtu-checkstyle and guava. If their violations are ok, I am ok to merge.
Good tool for site checkstyle site diff - https://github.com/attatrol/ahsm , example : http://attatrol.github.io/site_result/checkstyle_merged.html

**Comment by Vladlis** [comment]
@romani 
http://vladlis.github.io/reports/final_local_variable/site_result/
report looks OK, I haven't found any false-positives there

**Comment by romani** [comment]
merged as FF, thanks a lot for complicated update!

@Vladlis , reports are ok, you can send PR to Guava.

@attatrol , can you take a look why collors are missed in report.

**Comment by Vladlis** [comment]
@romani 
colors are missed, because I did not change the default stylesheet, I guess

**Comment by attatrol** [comment]
@Vladlis 
Css are included with the latest commit.

**Pull request closed without merging**